django/db/backends/mysql/schema.py

```diff
@@ -135,24 +153,63 @@ def _field_should_be_indexed(self, model, field):
             return False
         return not self._is_limited_data_type(field)
 
-    def _delete_composed_index(self, model, fields, *args):
+    def _create_missing_fk_index(
+        self,
+        model,
+        *,
+        fields,
+        expressions=None,
+    ):
         """
         MySQL can remove an implicit FK index on a field when that field is
         covered by another index like a unique_together. "covered" here means
         that the more complex index starts like the simpler one.
-        https://bugs.mysql.com/bug.php?id=37910 / Django ticket #24757
-        We check here before removing the [unique|index]_together if we have to
+        https://bugs.mysql.com/bug.php?id=37910 / Django tickets #24757,31335
+        We check here before removing the composed index if we have to
         recreate a FK index.
         """
-        first_field = model._meta.get_field(fields[0])
+
+        first_field_name = None
+        if fields:
+            if (
+                fields[0].startswith("-")
+                and not self.connection.features.supports_index_column_ordering
+            ):
+                first_field_name = fields[0][1:]
+            elif not fields[0].startswith("-"):
+                first_field_name = fields[0]
+        elif (
+            expressions
+            and self.connection.features.supports_expression_indexes
+            # expression is a simple F-object
+            and isinstance(expressions[0], F)
+            # expression doesn't contain custom lookup (field__lookup)
+            and LOOKUP_SEP not in expressions[0].name
+        ):
+            first_field_name = expressions[0].name
+
+        if not first_field_name:
+            return
+
+        meta = model._meta
+
+        first_field = meta.get_field(first_field_name)
```

Comment:
`meta` is used only here so a temporary variable is unnecessary:
```suggestion
        first_field = model._meta.get_field(first_field_name)
```